node/subscription.go

```diff
@@ -54,22 +51,96 @@ func (s *subscription) Unsubscribe(ctx context.Context) error {
 	}
 
 	if response.Error != nil {
-		return errors.Errorf("%v", response.Error)
+		return errors.Errorf("%s", string(*response.Error))
 	}
 
-	s.cancel()
 	return nil
 }
 
-func (s *subscription) Done() <-chan struct{} {
-	return s.ctx.Done()
+func newSubscription(response *jsonrpc.RawResponse, id string, r Requester) *subscription {
+	s := subscription{
+		response:        response,
+		subscriptionID:  id,
+		notificationsCh: make(chan *jsonrpc.Notification),
+		dispatchCh:      make(chan *jsonrpc.Notification),
+		signalCh:        make(chan struct{}),
+		stoppedCh:       make(chan struct{}),
+		conn:            r,
+	}
+
+	// start a goroutine to move work from dispatchCh to notificationsCh
+	// we have to have this intermediate channel and the two
+	// signalling channels to make sure that everything
+	// is shut down in the proper order and no one tries writing
+	// to a closed channel.  See for example variant #5 on https://go101.org/article/channel-closing.html
+	go func() {
+		for {
+			select {
+			case <-s.signalCh:
+				// we've been signalled to stop.  Close the stopped channel so it
+				// signals to everyone we've already been stopped.
+				close(s.stoppedCh)
+
+				// then close the notifications channel so any consumers of
+				// subscription.Ch() are unblocked
+				close(s.notificationsCh)
+			case n := <-s.dispatchCh:
+				// we have a notification to dispatch to the external channel
+				select {
+				case s.notificationsCh <- n:
+					// we successfully dispatched the notification
+				case <-s.signalCh:
+					// we were told to stop while trying to write the notification
+					// Close the stopped channel so it signals to everyone we've
+					// already been stopped.
+					close(s.stoppedCh)
+
```

Comment:
minor:

In the article reference above, they still do what amounts to a `s.notifications <- n` here, but to me that seemed superfluous as someone has already called `.stop()`.